python-sdk/nuscenes/predict/models/mtp.py

```diff
@@ -0,0 +1,325 @@
+        avg_distance = torch.mean(l2_norms)
+        return avg_distance
+
+    def _compute_angles_from_ground_truth(self, target: torch.Tensor,
+                                          trajectories: torch.Tensor) -> List[float]:
+        """
+        Compute angle between the target trajectory (ground truth) and
+        the predicted trajectories.
+        :param target: Shape [1, n_timesteps, 2]
+        :param trajectories: Shape [n_modes, n_timesteps, 2
+        :return: List of angles.
+        """
+        angles_from_ground_truth = []
+        for mode, mode_trajectory in enumerate(trajectories):
+            # For each mode, we compute the angle between the last point of the predicted trajectory for that
+            # mode and the last point of the ground truth trajectory.
+            angle = self._angle_between(target[0], mode_trajectory)
+
+            angles_from_ground_truth.append((angle, mode))
+        return angles_from_ground_truth
+
+    def _compute_best_mode(self,
+                           angles_from_ground_truth: List[float],
+                           target: torch.Tensor, trajectories: torch.Tensor) -> int:
+        """
+        Finds the index of the best mode given the angles from the ground truth.
+        :param angles_from_ground_truth: List of angles
+        :param target: Shape [1, n_timesteps, 2]
+        :param trajectories: Shape [n_modes, n_timesteps, 2]
+        """
+
+        # We first sort the modes based on the angle to the ground truth (ascending order), and keep track of
+        # the index corresponding to the biggest angle that is still smaller than a threshold value.
+        angles_from_ground_truth = sorted(angles_from_ground_truth)
+        max_angle_below_thresh_idx = -1
+        for angle_idx, (angle, mode) in enumerate(angles_from_ground_truth):
+            if angle <= self.angle_threshold:
+                max_angle_below_thresh_idx = angle_idx
+            else:
+                break
+
+        # We choose the best mode at random IF there are no modes with an angle less than the threshold.
+        if max_angle_below_thresh_idx == -1:
+            best_mode = random.randint(0, self.num_modes - 1)
+
+        # We choose the best mode to be the one that provides the lowest ave of l2 norms between the
+        # predicted trajectory and the ground truth, taking into account only the modes with an angle
+        # less than the threshold IF there is at least one mode with an angle less than the threshold.
+        else:
+            # Out of the selected modes above, we choose the final best mode as that which returns the
+            # smallest ave of l2 norms between the predicted and ground truth trajectories.
+            distances_from_ground_truth = []
+
+            for angle, mode in angles_from_ground_truth[:max_angle_below_thresh_idx + 1]:
+                norm = self._compute_ave_l2_norms(target - trajectories[mode, :, :]).item()
```

Comment:
`_compute_ave_l2_norms()` supposedly returns a float.
I suggest you move the `.item()` into that function.

Reply:
Good idea.